extensions/amp-access/0.1/amp-access.js

```diff
@@ -834,166 +611,37 @@ export class AccessService {
-        this.openLoginDialog_(loginUrl);
-    const loginPromise = dialogPromise.then(result => {
-      dev().fine(TAG, 'Login dialog completed: ', eventLabel, result);
-      this.loginPromise_ = null;
-      const query = parseQueryString(result);
-      const s = query['success'];
-      const success = (s == 'true' || s == 'yes' || s == '1');
-      if (success) {
-        this.loginAnalyticsEvent_(eventLabel, 'success');
-      } else {
-        this.loginAnalyticsEvent_(eventLabel, 'rejected');
-      }
-      const exchangePromise = this.signIn_.postLoginResult(query) ||
-          Promise.resolve();
-      if (success || !s) {
-        // In case of a success, repeat the authorization and pingback flows.
-        // Also do this for an empty response to avoid false negatives.
-        // Pingback is repeated in this case since this could now be a new
-        // "view" with a different access profile.
-        return exchangePromise.then(() => {
-          this.broadcastReauthorize_();
-          return this.runAuthorization_(/* disableFallback */ true)
-              .then(() => {
-                this.scheduleView_(/* timeToView */ 0);
-              });
-        });
-      }
-    }).catch(reason => {
-      dev().fine(TAG, 'Login dialog failed: ', eventLabel, reason);
-      this.loginAnalyticsEvent_(eventLabel, 'failed');
-      if (this.loginPromise_ == loginPromise) {
-        this.loginPromise_ = null;
-      }
-      throw reason;
-    });
-    this.loginPromise_ = loginPromise;
-    this.loginStartTime_ = now;
-    return this.loginPromise_;
+    // Namespaced
+    const namespace = type.substring(0, splitPoint);
+    const match = this.sources_.filter(s => s.getNamespace() == namespace);
+    user().assert(match.length, 'Login type not found: %s', namespace);
+    return match[0].loginWithType(type.substring(splitPoint + 1));
   }
 
   /**
-   * @param {string} type
-   * @param {string} event
-   * @private
+   * Either combine namespaced responses or just return the single one.
+   *
+   * @return {?JsonObject}
    */
-  loginAnalyticsEvent_(type, event) {
-    this.analyticsEvent_(`access-login-${event}`);
-    if (type) {
-      this.analyticsEvent_(`access-login-${type}-${event}`);
+  combinedResponses() {
+    if (this.sources_.length == 1 && !this.sources_[0].getNamespace()) {
+      return this.sources_[0].getAuthResponse();
     }
-  }
 
-  /**
-   * @return {?Promise<!Array<!{type: string, url: string}>>}
-   * @private
-   */
-  buildLoginUrls_() {
-    if (Object.keys(this.loginConfig_).length == 0) {
-      return null;
-    }
-    const promises = [];
-    for (const k in this.loginConfig_) {
-      promises.push(
-          this.buildUrl(this.loginConfig_[k], /* useAuthData */ true)
-              .then(url => {
-                this.loginUrlMap_[k] = url;
-                return {type: k, url};
-              }));
-    }
-    return Promise.all(promises);
+    const combined = /** @type {JsonObject} */ ({});
```

Comment:
`!JsonObject`

Reply:
this actually could be null if somehow called before the first reponse has arrived

Reply:
But you do initialize it here as const into non-null, looks like.

Reply:
the problem is that in the single-source case right above this we directly response getAuthReponse() from the source which defaults to null. it looks like it would be okay to make that default to an empty dictionary, but i wasn't sure if i was missing anything.

Reply:
(maybe it changes getAuthdataField semantics in some subtle way, although I don't think it does?)